lib/grp.gi

```diff
@@ -4291,86 +4300,145 @@ end );
+        triv:=true;
+        typename:="defaultFinitelyGeneratedTrivialGroupType";
       fi;
-      typ:=fam!.defaultGroupType;
+    elif Length(gens)<=1 then
+      typ:=typ and IsCyclic;
+      typename:="defaultFinitelyGeneratedCyclicGroupType";
     fi;
+  fi;
+  if not IsBound(fam!.(typename)) then
+    typ:=NewType(fam,typ);
+    fam!.(typename):=typ;
+  else
+    typ:=fam!.(typename);
+  fi;
 
-    G:=rec();
-    ObjectifyWithAttributes(G,typ,GeneratorsOfMagmaWithInverses,AsList(gens));
+  G:=rec();
+  if triv=fail then
+    ObjectifyWithAttributes(G,typ,GeneratorsOfMagmaWithInverses,AsList(gens),
+                            IsEmpty,false);
+  else
+    ObjectifyWithAttributes(G,typ,GeneratorsOfMagmaWithInverses,AsList(gens),
+                            IsEmpty,false,IsTrivial,triv,
+                            IsNonTrivial,not triv);
+  fi;
 
-    return G;
-    end );
+  return G;
+end );
 
 InstallMethod( GroupWithGenerators,
     "generic method for collection and identity element",
     IsCollsElms, [ IsCollection, IsMultiplicativeElementWithInverse ],
-    function( gens, id )
-    local G,fam,typ;
-
-    fam:=FamilyObj(gens);
-    if IsFinite(gens) then
-      if not IsBound(fam!.defaultFinitelyGeneratedGroupWithOneType) then
-        fam!.defaultFinitelyGeneratedGroupWithOneType:=
-          NewType(fam,IsGroup and IsAttributeStoringRep
-                      and HasGeneratorsOfMagmaWithInverses
-                      and IsFinitelyGeneratedGroup and HasOne);
-      fi;
-      typ:=fam!.defaultFinitelyGeneratedGroupWithOneType;
-    else
-      if not IsBound(fam!.defaultGroupWithOneType) then
-        fam!.defaultGroupWithOneType:=
-          NewType(fam,IsGroup and IsAttributeStoringRep
-                      and HasGeneratorsOfMagmaWithInverses and HasOne);
+function( gens, id )
+local G,fam,typ,triv,typename;
+
+  fam:=FamilyObj(gens);
+  triv:=fail; # might not find out
+
+  typ:=IsGroup and IsAttributeStoringRep 
+        and HasIsEmpty and HasGeneratorsOfMagmaWithInverses and HasOne;
+  typename:="defaultGroupWithOneType";
+
+  if IsFinite(gens) then
+    typ:=typ and IsFinitelyGeneratedGroup;
+    typename:="defaultFinitelyGeneratedGroupWithOneType";
+
+    if Length(gens)>0 and CanEasilyCompareElements(gens) then
+      if ForAny(gens,x->x<>id) then
+        typ:=typ and HasIsTrivial and HasIsNonTrivial;
+        triv:=false;
+        if Length(gens)<=1 then
+          typ:=typ and IsCyclic;
+          typename:="defaultFinitelyGeneratedNontrivialCyclicGroupWithOneType";
+        else
+          typename:="defaultFinitelyGeneratedNontrivialGroupWithOneType";
+        fi;
+      else
+        typ:=typ and HasIsTrivial and HasIsNonTrivial;
+        triv:=true;
+        typename:="defaultFinitelyGeneratedTrivialGroupWithOneType";
       fi;
-      typ:=fam!.defaultGroupWithOneType;
+    elif Length(gens)<=1 then
+      typ:=typ and IsCyclic;
+      typename:="defaultFinitelyGeneratedCyclicGroupWithOneType";
```

Comment:
This type computation is repeated several times across multiple GroupWithGenerators methods - can't it perhaps be factored out into a common helper function?

Reply:
The best I was able to come up with would have been preprocessor macros, but for a function there are too many small differences that would require input parameter and extra processing.  (This is an observation of my incompetence, not a statement that it cannot be done.)